File: Task1C/task1c/src/main/java/com/cps2004/VideoGame.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public abstract class VideoGame extends Product{
    private String platform;
    private String developer;

    // Default constructor
    public VideoGame(){
        super();
        this.platform = "";
        this.developer = "";
    }

    // Parameterized constructor
    public VideoGame(String name, int id, Supplier supplier, int price, int volume, String platform, String developer){
        super(name, id, supplier, price, volume);
        this.platform = platform;
        this.developer = developer;
    }

    // Getters and setters
    public String getPlatform() { return platform; }
    public String getDeveloper() { return developer; }
    public void setPlatform(String platform) { this.platform = platform; }
    public void setDeveloper(String developer) { this.developer = developer; }

    @Override
    // Calculate discounted price
    public abstract double calculate_discount(String month, int quantity);

    // Display product information
    public void display_product_info() {
        super.display_product_info();
        System.out.println("Product platform: " + platform);
        System.out.println("Product developer: " + developer);
    }

    // Save product information to file
    public void save(ObjectNode newItem) {
        super.save(newItem);
        newItem.put("platform", this.platform);
        newItem.put("developer", this.developer);
    }

    // Load product information from file
    public void load(JsonNode node) {
        super.load(node);
        platform = node.get("platform").asText();
        developer = node.get("developer").asText();
    }
}

File: Task3C/videogame.cpp
#include "videogame.h"

// Default constructor
VideoGame::VideoGame() {
    this->platform = "";
    this->developer = "";
}

// Parameterised constructor
VideoGame::VideoGame(string name, int price, Supplier* supplier, int volume, int productID, string platform, string developer) : Product(name, price, supplier, volume, productID) {
    this->platform = platform;
    this->developer = developer;
}

// Getters and setters
string VideoGame::getPlatform() {
    return this->platform;
}

string VideoGame::getDeveloper() {
    return this->developer;
}

void VideoGame::setPlatform(string platform) {
    this->platform = platform;
}

void VideoGame::setDeveloper(string developer) {
    this->developer = developer;
}

// Display product information
void VideoGame::display_product_info() {
    Product::display_product_info();
    cout << "Platform: " << this->platform << endl;
    cout << "Developer: " << this->developer << endl;
}

// Save product information to file
void VideoGame::save(ostream& out) const {
    Product::save(out);
    out << platform << "," << developer << "\n";
}

// Load product information from file
void VideoGame::load(istream& in) {
    Product::load(in);

    string line;
    getline(in, line, '\n');
    stringstream ss(line);

    string token;
    getline(ss, token, ',');
    platform = token;
    getline(ss, token, ',');
    developer = token;
}